drivers/spi/spi_litespi.h

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright (c) 2019 Antmicro <www.antmicro.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef _SPI_LITESPI__H
+#define _SPI_LITESPI__H
+
+#include "spi_context.h"
+
+#include <sys/sys_io.h>
+#include <device.h>
+#include <drivers/spi.h>
+
+#define SPI_BASE_ADDR DT_INST_0_LITEX_SPI0_BASE_ADDRESS
+#define SPI_CONTROL_ADDR SPI_BASE_ADDR
+#define SPI_STATUS_ADDR (SPI_BASE_ADDR + 0x08)
+#define SPI_MOSI_DATA_ADDR (SPI_BASE_ADDR + 0x0c)
+#define SPI_MISO_DATA_ADDR (SPI_BASE_ADDR + 0x10)
+#define SPI_CS_ADDR (SPI_BASE_ADDR + 0x14)
+#define SPI_LOOPBACK_ADDR (SPI_BASE_ADDR + 0x18)
+
+#define POSITION_WORD_SIZE 8
+#define SPI_MAX_CS_SIZE 0x100
+#define SPI_WORD_SIZE 8
+
+#define SPI_ENABLE 0x1
+
+#define SPI_DATA(dev) ((struct spi_litespi_data *) ((dev)->driver_data))
+
+/* Function Declarations */
+
+static inline void csr_writel(u32_t value, unsigned long addr)
+{
+	*((volatile u32_t *)addr) = value;
+}
+
+static inline u32_t csr_readl(unsigned long addr)
+{
+	return *(volatile u32_t *)addr;
+}
+
+static inline unsigned char read8(unsigned long addr)
+{
+	return csr_readl(addr);
+}
+
+static inline void write8(unsigned char value, unsigned long addr)
+{
+	csr_writel(value, addr);
+}
+
+static inline void write16(unsigned short int value, unsigned long addr)
+{
+	csr_writel(value >> 8, addr);
+	csr_writel(value, addr + 0x4);
+}
```

Comment:
looks like riscv arch is missing sys_io.h implementation. Could you do that? it will certainly be useful for other part of the system on that architecture. And you would not have to re-implement it here for instance.

Reply:
We could definiately use `sys_read8`/`sys_write8` from `common/sys_io.h` and drop `csr_writel`/`csr_readl`/`read8`/`write8`.

As of `write16`, the situation is a bit more complicated. The way we handle 16-bit registers here is not RISC-V specific, but rather LiteX specific. LiteX is an SoC builder that can be configured in many ways (which influences the way `write16` should behave - it depends on `data_width` and `bus_width` parameters) and that can target different CPUs (so this driver is not limited to RISC-V cores). As a result I'm not sure if `sys_io.h` is the place to put this code.

I do agree, though, that it would be good to have a `litex_common.h` header that provides implementation of `read16`/`write16`/`read32`/`write32` functions for all LiteX drivers. Do you have a suggestion where should such header be put? Is simply `include/litex_common.h` (or `include/litex.h`) a proper place?

Reply:
We could definiately use `sys_read8`/`sys_write8` from `common/sys_io.h` and drop `csr_writel`/`csr_readl`/`read8`/`write8`.

As of `write16`, the situation is a bit more complicated. The way we handle 16-bit registers here is not RISC-V specific, but rather LiteX specific. LiteX is an SoC builder that can be configured in many ways (which influences the way `write16` should behave - it depends on `data_width` and `bus_width` parameters) and that can target different CPUs (so this driver is not limited to RISC-V cores). As a result I'm not sure if `sys_io.h` is the place to put this code.

I do agree, though, that it would be good to have a `litex_common.h` header that provides implementation of `read16`/`write16`/`read32`/`write32` functions for all LiteX drivers. Do you have a suggestion where should such header be put? Is simply `include/litex_common.h` (or just `include/litex.h`) a proper place?

Reply:
include/ is for generic headers. Instead, I would embed the specific functions into soc/riscv/litex-vexriscv/soc.h

Reply:
Sure. For now it's good enough to put it into `soc/riscv/litex-vexriscv`. In the future, if more LiteX-based platforms are added this should be moved to a more generic location.